### ISSUE TITLE
move to TypeScript 3.7

### ISSUE THREAD

**Issue by vankop** [stylelint/stylelint#4399]
There is a killer feature for stylelint - now it possible to create declaration file from JS https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#--declaration-and---allowjs

Blocked by:  https://github.com/stylelint/stylelint/issues/4245

**Comment by hudochenkov** [comment]
I don't think we need a separate issue for this. Upgrade would be handled as with any other dependency.

**Comment by vankop** [comment]
This issue was not about upgrade at first place, It is about creating definitions =)

**Comment by hudochenkov** [comment]
Then rename it and explain a little bit more in the first comment :)

**Comment by jeddy3** [comment]
This is neat. It's great to see JSDocs TypeScript getting some love.

**Comment by jeddy3** [comment]
With #4245 now closed, I'm removing the '"blocked" label.

**Comment by hudochenkov** [comment]
Currently it's blocked by https://github.com/stylelint/stylelint/pull/4505

**Comment by ybiquitous** [comment]
> Currently it's blocked by #4505

We are using TypeScript 3.7 now,

https://github.com/stylelint/stylelint/blob/c63a45fbbde88010cd43703feb6591475db213ab/package.json#L116

and we will be able to use TypeScript 3.8 soon via #4626! 😄

**Comment by tyankatsu0105** [comment]
Hi.
I have a question.
I just found [types](https://github.com/stylelint/stylelint/tree/master/types).
But what for?
At least currently, the types have not been released.

**Comment by hudochenkov** [comment]
@tyankatsu0105 we use TypeScript to avoid bugs in our code https://github.com/stylelint/stylelint/blob/6cfea296736d3189a60eda91bb83aece26bd51c2/package.json#L44 We don't use TypeScript language, but JSDoc, which is supported by TypeScript type checker.

**Comment by tyankatsu0105** [comment]
@hudochenkov 
I see. So these types are not for creating stylelint plugin or something with stylelint but for supporting to develop stylelint self, right?

**Comment by hudochenkov** [comment]
Right. But we want to expose them in future, so plugin developers could use them. That's why we have this issue.

**Comment by tyankatsu0105** [comment]
>But we want to expose them in future, so plugin developers could use them

Awesome!!!
I created stylelint plugin with TypeScript.
https://github.com/tyankatsu0105/stylelint-typescript.
I was worried because @types/stylelint seemed like this wasn't already being maintained.

Thanks for the reply @hudochenkov :)

**Comment by glen-84** [comment]
I tried this now, and most of the declaration files appear to be generated, but there are at least two issues:

**`#1`**

> lib/utils/parseCalcExpression/parser.js:383:13 - error TS9005: Declaration emit for this file requires using private name 'Parser'. An explicit type annotation may unblock declaration emit.
> 
> 383             var parser = (function () {
>                 ~~~
> 
> 
> Found 1 error.

This looks like a generated file, so it probably can't be edited.

Possibly related to https://github.com/microsoft/TypeScript/issues/37832 or https://github.com/microsoft/TypeScript/issues/38382.

**`#2`**

Multiple declaration files have code like this:

```ts
export = _exports;
```

... with the error: `An export assignment cannot be used in a module with other exported elements.`

I couldn't find an existing TypeScript issue regarding this error, but [other projects](https://github.com/puppeteer/puppeteer/issues/5168) have run into this as well it seems.

---

You can test it yourself by adding the following to `tsconfig.json`:

```json
        "declaration": true,
	"emitDeclarationOnly": true,
	"outDir": "types-new"
```

(and optionally removing `"noEmit": true`, which is not needed)

**Comment by Phillip9587** [comment]
Any progress on this? It really sucks that there are only outdated types available. I may help if someone describes the current status to me...

**Comment by jeddy3** [comment]
> I may help if someone describes the current status to me...

The last attempt (https://github.com/stylelint/stylelint/issues/4399#issuecomment-650538145) to do this was a year ago and some issues were encountered. Things may have improved since then. So, if you've time you're welcome to look into it. It would be amazing to resolve as TypeScript usage has exploded.

**Comment by glen-84** [comment]
Apart from the 2 issues previously mentioned (which still exist – perhaps related to https://github.com/microsoft/TypeScript/issues/41672), there is also an issue in `lib/formatters/index.js`, where there is a missing argument (x6).

```
lib/formatters/index.js:6:11 - error TS2554: Expected 1 arguments, but got 0.

6  compact: importLazy(() => require('./compactFormatter'))(),
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/import-lazy/index.d.ts:24:5
    24 ): (moduleId: string) => T;
           ~~~~~~~~~~~~~~~~
    An argument for 'moduleId' was not provided.
```

**Comment by Phillip9587** [comment]
@jeddy3 @glen-84 Would migrating Stylelint to Typescript be an option or are there any reasons you haven't done that yet?

**Comment by ybiquitous** [comment]
@glen-84 About <https://github.com/stylelint/stylelint/issues/4399#issuecomment-803565238>, I've opened the issue to `import-lazy`:
https://github.com/sindresorhus/import-lazy/issues/19

But, to fix the TS error, it seems that we could just specify a module ID as below:

```diff
-	compact: importLazy(() => require('./compactFormatter'))(),
+	compact: importLazy(() => require('./compactFormatter'))('compactFormatter'),
```

**Comment by glen-84** [comment]
@ybiquitous Is stylelint even being bundled in the first place?

**Comment by ybiquitous** [comment]
@glen-84 Do you mean stylelint provides a bundle version for browsers?
If so, the answer perhaps is “no” (see also #4796).

I think this lazy loading is for performance, as I see #4278.

**Comment by glen-84** [comment]
> @glen-84 Do you mean stylelint provides a bundle version for browsers?

Yes. It looks like @jeddy3 changed the code [here](https://github.com/stylelint/stylelint/commit/160c6ae1af03640141754f875731822683297df1).

It could probably be reverted to the standard usage of `import-lazy` until a decision is made WRT bundling, but there are other issues with the TypeScript build anyway.

**Comment by marcfallows** [comment]
Given types exist in https://github.com/stylelint/stylelint/tree/master/types/stylelint, couldn't these be exposed in the `files` in package.json so consumers of the npm package can use them?
